src/crab/type_domain.cpp

```diff
@@ -0,0 +1,714 @@
+    if (is_bottom()) {
+        return;
+    }
+    m_vars[var] = nullptr;
+}
+
+void register_types_t::set_to_bottom() {
+    m_vars = live_registers_t{nullptr};
+    m_is_bottom = true;
+}
+
+void register_types_t::set_to_top() {
+    m_vars = live_registers_t{nullptr};
+    m_is_bottom = false;
+}
+
+bool register_types_t::is_bottom() const { return m_is_bottom; }
+
+bool register_types_t::is_top() const {
+    if (m_is_bottom) { return false; }
+    if (m_all_types == nullptr) return true;
+    for (auto it : m_vars) {
+        if (it != nullptr) return false;
+    }
+    return true;
+}
+
+void register_types_t::insert(register_t reg, const reg_with_loc_t& reg_with_loc, const ptr_t& type) {
+    (*m_all_types)[reg_with_loc] = type;
+    m_vars[reg] = std::make_shared<reg_with_loc_t>(reg_with_loc);
+}
+
+std::optional<ptr_t> register_types_t::find(reg_with_loc_t reg) const {
+    auto it = m_all_types->find(reg);
+    if (it == m_all_types->end()) return {};
+    return it->second;
+}
+
+std::optional<ptr_t> register_types_t::find(register_t key) const {
+    if (m_vars[key] == nullptr) return {};
+    const reg_with_loc_t& reg = *(m_vars[key]);
+    return find(reg);
+}
+
+stack_t stack_t::operator|(const stack_t& other) const {
+    if (is_bottom() || other.is_top()) {
+        return other;
+    } else if (other.is_bottom() || is_top()) {
+        return *this;
+    }
+    offset_to_ptr_t out_ptrs;
+    for (auto const&kv: m_ptrs) {
+        auto it = other.find(kv.first);
+        if (it && kv.second == it.value())
+            out_ptrs.insert(kv);
+    }
+    return stack_t(std::move(out_ptrs), false);
+}
+
+void stack_t::operator-=(int key) {
+    auto it = find(key);
+    if (it)
+        m_ptrs.erase(key);
+}
+
+void stack_t::set_to_bottom() {
+    m_ptrs.clear();
+    m_is_bottom = true;
+}
+
+void stack_t::set_to_top() {
+    m_ptrs.clear();
+    m_is_bottom = false;
+}
+
+stack_t stack_t::bottom() { return stack_t(true); }
+
+stack_t stack_t::top() { return stack_t(false); }
+
+bool stack_t::is_bottom() const { return m_is_bottom; }
+
+bool stack_t::is_top() const {
+    if (m_is_bottom)
+        return false;
+    return m_ptrs.empty();
+}
+
+void stack_t::insert(int key, ptr_t value) {
+    m_ptrs[key] = value;
+}
+
+std::optional<ptr_t> stack_t::find(int key) const {
+    auto it = m_ptrs.find(key);
+    if (it == m_ptrs.end()) return {};
+    return it->second;
+}
+
+}
+
+bool type_domain_t::is_bottom() const {
+    return (m_stack.is_bottom() || m_types.is_bottom());
+}
+
+bool type_domain_t::is_top() const {
+    return (m_stack.is_top() && m_types.is_top());
+}
+
+type_domain_t type_domain_t::bottom() {
+    type_domain_t typ;
+    typ.set_to_bottom();
+    return typ;
+}
+
+void type_domain_t::set_to_bottom() {
+    m_types.set_to_bottom();
+}
+
+void type_domain_t::set_to_top() {
+    m_stack.set_to_top();
+    m_types.set_to_top();
+}
+
+bool type_domain_t::operator<=(const type_domain_t& abs) const {
+    return true;
```

Comment:
Write a big warning here that the operation is not implemented. This operation is very important if they program has loops.

Reply:
ok